src/Microsoft.DotNet.Wpf/src/PresentationCore/MS/internal/Text/TextInterface/Factory.cs

```diff
@@ -0,0 +1,342 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Runtime.InteropServices;
+using MS.Internal.Interop;
+using MS.Internal.Interop.DWrite;
+using MS.Internal.Text.TextInterface.Interfaces;
+
+namespace MS.Internal.Text.TextInterface
+{
+    /// <summary>
+    /// The root factory interface for all DWrite objects.
+    /// </summary>
+    internal unsafe class Factory
+    {
+        private const int DWRITE_E_FILEFORMAT = unchecked((int)0x88985000L);
+
+        // b859ee5a-d838-4b5b-a2e8-1adc7d93db48
+        private static readonly Guid IID_IDWriteFactory = new Guid(0xb859ee5a, 0xd838, 0x4b5b, 0xa2, 0xe8, 0x1a, 0xdc, 0x7d, 0x93, 0xdb, 0x48);
+
+        /// <summary>
+        /// A pointer to the wrapped DWrite factory object.
+        /// </summary>
+        private NativeIUnknownWrapper<IDWriteFactory> _factory;
+
+        /// <summary>
+        /// The custom loader used by WPF to load font files.
+        /// </summary>
+        private readonly FontFileLoader _wpfFontFileLoader;
+
+        /// <summary>
+        /// The custom loader used by WPF to load font collections.
+        /// </summary>
+        private readonly FontCollectionLoader _wpfFontCollectionLoader;
+
+        private readonly IFontSourceFactory _fontSourceFactory;
+
+        /// <summary>
+        /// Constructs a factory object.
+        /// </summary>
+        /// <param name="factoryType">Identifies whether the factory object will be shared or isolated.</param>
+        /// <param name="fontSourceCollectionFactory">A factory object that will create managed FontSourceCollection 
+        /// objects that will be utilized to load embedded fonts.</param>
+        /// <param name="fontSourceFactory">A factory object that will create managed FontSource
+        /// objects that will be utilized to load embedded fonts.</param>
+        /// <returns>
+        /// The factory just created.
+        /// </returns>
+        private Factory(FactoryType factoryType, IFontSourceCollectionFactory fontSourceCollectionFactory, IFontSourceFactory fontSourceFactory)
+        {
+            Initialize(factoryType);
+
+            _wpfFontFileLoader = new FontFileLoader(fontSourceFactory);
+            _wpfFontCollectionLoader = new FontCollectionLoader(
+                                                               fontSourceCollectionFactory,
+                                                               _wpfFontFileLoader
+                                                               );
+
+            _fontSourceFactory = fontSourceFactory;
+
+            IntPtr pIDWriteFontFileLoaderMirror = Marshal.GetComInterfaceForObject(
+                                                    _wpfFontFileLoader,
+                                                    typeof(IDWriteFontFileLoaderMirror));
```

Comment:
I've some doubts about a few of the low-level COM work here, not because it is wrong but because I think relying on built-in COM interop first is safer for the port and then it can be revisited. The counter is NativeAOT, but the above line is not AOT friendly. I think it might make more sense for the first pass to be reliant on the built-in COM interop and then go back through and make it AOT friendly. It will take longer but higher chances of success.

Reply:
I support any progress even if it’s not immediate and via some checkpoints. But that’s not my PR but @ThomasGoulet73 so let he speak his mind.

Reply:
Okay. Let's see how this goes. Converting C++ directly to C# isn't all that good of an idea in general because it tends to generate poor and unsafe C#. It should be a priority to change this prior to moving onto the next section though.